File: db4on/test/src/com/db4o/test/BindFileSize.cs
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

using System;
using System.Text;

namespace com.db4o.test {

   public class BindFileSize {
      static internal int LENGTH = 10000;
      internal String foo;
      
      public BindFileSize() : base() {
      }
      
      public BindFileSize(int Length) : base() {
         StringBuilder sb1 = new StringBuilder();
         for (int i1 = 0; i1 < Length; i1++) {
            sb1.Append("g");
         }
         this.foo = sb1.ToString();
      }
      
      public void Store() {
         Tester.DeleteAllInstances(this);
         Tester.Store(new BindFileSize(LENGTH));
      }
      
      public void TestGrowth() {
         int call1 = 0;
         BindFileSize bfs1 = (BindFileSize)Tester.GetOne(this);
         long id1 = Tester.ObjectContainer().GetID(bfs1);
         for (int i1 = 0; i1 < 12; i1++) {
            bfs1 = new BindFileSize(LENGTH);
            Tester.ObjectContainer().Bind(bfs1, id1);
            Tester.ObjectContainer().Set(bfs1);
            Tester.Commit();
            CheckFileSize(call1++);
            Tester.ReOpen();
         }
      }
      
      private void CheckFileSize(int call) {
         if (Tester.CanCheckFileSize()) {
            int newFileLength1 = Tester.FileLength();
            if (call == 6) {
               jumps = 0;
               fileLength = newFileLength1;
            } else if (call > 6) {
               if (newFileLength1 > fileLength) {
                  if (jumps < 4) {
                     fileLength = newFileLength1;
                     jumps++;
                  } else {
                     Tester.Error();
                  }
               }
            }
         }
      }
      [Transient] private static int fileLength;
      [Transient] private static int jumps;
   }
}

File: db4oj.tests/src/com/db4o/test/BindFileSize.java
/* Copyright (C) 2004   db4objects Inc.   http://www.db4o.com */

package com.db4o.test;

public class BindFileSize {
	
	static final int LENGTH = 10000;
	
	String foo;
	
	public BindFileSize(){
	}
	
	public BindFileSize(int length){
		StringBuffer sb = new StringBuffer(length);
		for (int i = 0; i < length; i++) {
            sb.append("g");
        }  
		this.foo = sb.toString();
	}
    
	public void store(){
		Test.deleteAllInstances(this);
		Test.store(new BindFileSize(LENGTH));
	}
	
	public void testGrowth(){
		int call = 0;
        
        Test.reOpen();
		
		BindFileSize bfs =  (BindFileSize)Test.getOne(this);
		long id = Test.objectContainer().getID(bfs);
		for (int i = 0; i < 12; i++) {
			bfs = new BindFileSize(LENGTH);
			Test.objectContainer().bind(bfs, id);
			Test.objectContainer().set(bfs);
			Test.commit();
			checkFileSize(call++);
			Test.reOpen();
		}
	}
	
	private void checkFileSize(int call){
		if(Test.canCheckFileSize()){
			int newFileLength = Test.fileLength();
			
			// Interesting for manual tests:
			// System.out.println(newFileLength);
			
			if(call == 6){
				// consistency reached, start testing
				jumps = 0;
				fileLength = newFileLength;
			}else if(call > 6){
				if(newFileLength > fileLength){
					if(jumps < 4){
						fileLength = newFileLength;
						jumps ++;
						// allow two further steps in size
						// may be necessary for commit space extension
					}else{
						// now we want constant behaviour
						Test.error();
					}
				}
			}
		}
	}
	
	private static transient int fileLength;
	private static transient int jumps; 



}
